macros/src/lib.rs

```diff
@@ -277,3 +279,117 @@ pub fn modal(input: TokenStream) -> TokenStream {
+both of which will have the `slash_command` and `prefix_command` attributes.
+
+`command_one` will have the default user_cooldown of 1000, while `command_two` overrides it to 2000.
+
+```rust
+struct MyCommands;
+
+#[poise::group(slash_command, prefix_command, user_cooldown=1000)]
+impl MyCommands {
+    /// This is a command
+    #[poise::command()]
+    async fn command_one(ctx: Context<'_>) -> Result<(), Error> {
+        // code
+    }
+
+    /// This is another command
+    #[poise::command(user_cooldown=2000)]
+    async fn command_two(ctx: Context<'_>) -> Result<(), Error> {
+        // code
+    }
+}
+```
+*/
+#[proc_macro_attribute]
+pub fn group(args: TokenStream, input_item: TokenStream) -> TokenStream {
+    match group_impl(args, input_item) {
+        Ok(x) => x,
+        Err(err) => err.write_errors().into(),
+    }
+}
+
+fn group_impl(args: TokenStream, input_item: TokenStream) -> Result<TokenStream, Error> {
+    let args = darling::ast::NestedMeta::parse_meta_list(args.into())?;
+
+    let group_args = <command::GroupArgs as darling::FromMeta>::from_list(&args)?;
+
+    // let item_impl = syn::parse_macro_input!(input_item as syn::ItemImpl);
+    let item_impl = match syn::parse::<syn::ItemImpl>(input_item) {
+        Ok(syntax_tree) => syntax_tree,
+        Err(err) => return Err(err.into()),
+    };
+    let name = item_impl.self_ty;
+
+    // vector of all #[poise::command(...)] command idents
+    let mut command_idents = vec![];
+
+    // collect each ImplItem in a stream
+    let mut impl_body = quote!();
+
+    for item in item_impl.items.iter() {
+        let mut item_stream = quote!(#item);
+
+        // if it's a function...
+        if let syn::ImplItem::Fn(f) = item {
+            // ... and it's a command
+            if let Some(attr) = f.attrs.iter().find(|attr| is_command_attr(attr)) {
+                // add to command list
+                command_idents.push(f.sig.ident.clone());
+
+                // Turn a syn::Attribute into command::CommandArgs
+                let attr_args = &attr.meta.require_list()?.tokens;
+
+                let command_args = darling::ast::NestedMeta::parse_meta_list(quote! {#attr_args})?;
+                let command_args =
+                    <command::CommandArgs as darling::FromMeta>::from_list(&command_args)?;
+
+                let new_args = command_args.from_group_args(&group_args);
+                let function = syn::ItemFn {
+                    attrs: vec![],
+                    vis: f.vis.clone(),
+                    sig: f.sig.clone(),
+                    block: Box::new(f.block.clone()),
+                };
+                item_stream = command::command(new_args, function)?.into();
+            }
+        }
+        impl_body = quote!(
+            #impl_body
+            #item_stream
+        );
+    }
+
+    Ok(quote! {
+        impl #name {
+            fn commands() -> Vec<Command<Data, Error>> {
+                vec![#(#name::#command_idents()),*]
```

Comment:
Nitpick, feel free to ignore: This could be a fixed length array, avoiding the heap alloc for minimal performance on startup which really doesn't matter.